pkg/controller/operconfig/migration.go

```diff
@@ -0,0 +1,80 @@
+package operconfig
+
+import (
+	"context"
+	"log"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	uns "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	operv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/cluster-network-operator/pkg/apply"
+	cnoclient "github.com/openshift/cluster-network-operator/pkg/client"
+)
+
+var gvrEgressFirewall = schema.GroupVersionResource{Group: "k8s.ovn.org", Version: "v1", Resource: "egressfirewalls"}
+var gvrEgressNetworkPolicy = schema.GroupVersionResource{Group: "network.openshift.io", Version: "v1", Resource: "egressnetworkpolicies"}
+
+func migrateEgressFirewallCRs(ctx context.Context, operConfig *operv1.Network, client cnoclient.Client) error {
+	switch operConfig.Spec.DefaultNetwork.Type {
+	case operv1.NetworkTypeOpenShiftSDN:
+		return convertEgressNetworkPolicyToEgressFirewall(ctx, client)
+	case operv1.NetworkTypeOVNKubernetes:
+		return convertEgressFirewallToEgressNetworkPolicy(ctx, client)
+	}
+
+	return nil
+}
+
+func convertEgressNetworkPolicyToEgressFirewall(ctx context.Context, client cnoclient.Client) error {
+	enpList, err := client.Default().Dynamic().Resource(gvrEgressNetworkPolicy).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	for _, enp := range enpList.Items {
+		log.Printf("Convert EgressNetworkPolicy %s/%s", enp.GetNamespace(), enp.GetName())
+
+		egressFirewall := &uns.Unstructured{
+			Object: map[string]interface{}{
+				"apiVersion": "k8s.ovn.org/v1",
+				"kind":       "EgressFirewall",
+				"metadata": map[string]interface{}{
+					"name":      enp.GetName(),
+					"namespace": enp.GetNamespace(),
+				},
+				"spec": enp.Object["spec"],
+			},
+		}
+		if err := apply.ApplyObject(ctx, client, egressFirewall, ""); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func convertEgressFirewallToEgressNetworkPolicy(ctx context.Context, client cnoclient.Client) error {
+	efList, err := client.Default().Dynamic().Resource(gvrEgressFirewall).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	for _, ef := range efList.Items {
+		log.Printf("Convert EgressNetworkPolicy %s/%s", ef.GetNamespace(), ef.GetName())
+
+		egressNetworkPolicy := &uns.Unstructured{
+			Object: map[string]interface{}{
+				"apiVersion": "network.openshift.io/v1",
+				"kind":       "EgressNetworkPolicy",
+				"metadata": map[string]interface{}{
+					"name":      ef.GetName(),
+					"namespace": ef.GetNamespace(),
+				},
+				"spec": ef.Object["spec"],
```

Comment:
If I am reading the spec right, egress firewall rule may have `ports` section https://docs.openshift.com/container-platform/4.11/networking/ovn_kubernetes_network_provider/configuring-egress-firewall-ovn.html#egressnetworkpolicy-rules_configuring-egress-firewall-ovn, while egressnetworkpolicy doesn't have this field https://docs.openshift.com/container-platform/4.11/networking/openshift_sdn/configuring-egress-firewall.html#egressnetworkpolicy-rules_openshift-sdn-egress-firewall
it means that we can properly migrate from ovn to sdn only if this field is not set, or e.g. just drop ports section and log big warning

Reply:
That is not our target. We only handle the migration from SDN to OVN and the rollback. So we only need to ensure that all the SDN config can be converted to OVN, not vice versa. Users shall not add unsupported egress firewall configuration before rolling back.

Reply:
So to understand better how this is supposed to work, users should ensure that the configuration is compatible or else can't migrate? What is an `unsupported egress firewall configuration`? Are we missing some context here that can help us review?

Reply:
What users have in Openshift SDN can be migrated to OVNKubernetes. And if they find some issue after the migration, they shall be able to roll back to SDN with the same configuration. So we don't expect, a user to do the opposite migration for a cluster with OVN at install time. So any fields supported only by OVN are out of the scope of the feature migration.

Reply:
The answer to this is that the user is not expected to make any changes during migration or within the time window where it decides whether the rollback has been successful or not. So basically they should not make any change in egress firewalls that are not supported in egress network policies until migration is done.
Egress firewall should be spec schema compatible with egress network policy.